### PR TITLE
Esp32s2: Expose more network parameters

### PR DESCRIPTION
**Pull request by astrobokonon**
When connected to an access point, this adds the following (super helpful, for me at least) information to ```wifi.radio```:

- ```ap_rssi```: RSSI
- ```ap_ssid```: SSID
- ```ap_bssid```: BSSID (usually it's MAC address)
- ```ipv4_gateway```: Gateway
- ```ipv4_subnet```: Subnet
- ```ipv4_dns```: Primary DNS

When not connected, these all just return ```None```.  I made sure RSSI is always the current one while walking around closer and farther from my AP in case someone wants to build a WiFi mapping toy robot, but I use it primarily for monitoring the current connection.

### CODE REVIEW

**Comment by astrobokonon** [comment]
Forgot to add BSSID (and maybe regular SSID) so those are up next, and then this would be ready for review.

**Comment by anecdata** [comment]
This will be very useful! Encryption type is another element commonly returned by Arduino and CircuitPython ESP32SPI.

File: shared-bindings/wifi/Radio.c [hunk 1]
```diff
@@ -163,6 +163,86 @@ STATIC mp_obj_t wifi_radio_connect(size_t n_args, const mp_obj_t *pos_args, mp_m
...
+MP_DEFINE_CONST_FUN_OBJ_1(wifi_radio_get_ap_rssi_obj, wifi_radio_get_ap_rssi);
+
+const mp_obj_property_t wifi_radio_ap_rssi_obj = {
+    .base.type = &mp_type_property,
+    .proxy = { (mp_obj_t)&wifi_radio_get_ap_rssi_obj,
+               (mp_obj_t)&mp_const_none_obj,
+               (mp_obj_t)&mp_const_none_obj },
+};
+
+//|     ap_ssid: int
```
Comment (tannewt) [inline shared-bindings/wifi/Radio.c:182]: What do you think about re-using the Network object to represent the current network? I think it has all of the `ap_` attributes you want here. Then you can just return None or a Network object.

File: shared-bindings/wifi/Radio.c [hunk 2]
```diff
@@ -163,6 +163,86 @@ STATIC mp_obj_t wifi_radio_connect(size_t n_args, const mp_obj_t *pos_args, mp_m
...
+MP_DEFINE_CONST_FUN_OBJ_1(wifi_radio_get_ap_bssid_obj, wifi_radio_get_ap_bssid);
+
+const mp_obj_property_t wifi_radio_ap_bssid_obj = {
+    .base.type = &mp_type_property,
+    .proxy = { (mp_obj_t)&wifi_radio_get_ap_bssid_obj,
+               (mp_obj_t)&mp_const_none_obj,
+               (mp_obj_t)&mp_const_none_obj },
+};
+
+//|     ipv4_gateway: Optional[ipaddress.IPv4Address]
```
Comment (tannewt) [inline shared-bindings/wifi/Radio.c:214]: Do these three properties need to be ipv4 specific? Could they just be `gateway`, `subnet` and `dns` and either return IPv4 or IPv6 addresses? Maybe we'd need to return more than one object then?

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for adding these! Just a couple ideas/comments about the API. Code looks good, just want to make sure it's the API we want.

File: shared-bindings/wifi/Radio.c (hunk 2 above)
Comment (astrobokonon) [reply shared-bindings/wifi/Radio.c:214]: Gateway and subnet are coming from the same ```esp_netif_ip_info_t``` that you got the ip address from so they're always going to be ipv4.  I just matched the naming scheme you used for the existing ```ipv4_address``` property honestly - it's easy enough to drop the prefixes, but I do think it's nice when everything was grouped together when you dir() the ```wifi.radio``` object.

For DNS, though, esp-idf's DNS type (```esp_netif_dns_info_t```) contains both ```esp_ip6_addr_t ip6``` and ```esp_ip4_addr_t```, so I was trying to be a little more clear on which one I was picking out.

File: shared-bindings/wifi/Radio.c (hunk 1 above)
Comment (astrobokonon) [reply shared-bindings/wifi/Radio.c:182]: I think you're right, I'll try that.  That'd avoid pulling out everything individually and would probably be nicer to interact with too.

File: shared-bindings/wifi/Radio.c (hunk 2 above)
Comment (tannewt) [reply shared-bindings/wifi/Radio.c:214]: Ok, I think the prefix is fine since it's possible to have both ipv4 and ipv6 versions of these.

**Comment by tannewt** [comment]
@astrobokonon Any update on this?

**Comment by astrobokonon** [comment]
@tannewt I haven't had a chance to circle back just yet, but I'm planning on carving out some time during the rest of this week/weekend!  Still would like to get this done for sure.

File: shared-bindings/wifi/Radio.c (hunk 1 above)
Comment (pvinci) [reply shared-bindings/wifi/Radio.c:182]: The current network is usually represented as an ip_interface isn't it? See https://docs.python.org/3/library/ipaddress.html#ipaddress.ip_interface 
The network object has the host bits and'ed away.

**Comment by tannewt** [comment]
Awesome! Thanks @astrobokonon

**Comment by astrobokonon** [comment]
@tannewt Ok, I think this is ready again.  Just tested it and it worked as before, and it is indeed nicer and cleaner.  

I admit that makes me sorta want to do the same with all the ```ipv4_``` prefixed properties to keep things more contained, but I'm a little hazy on how to best do that with the removal/depreciation of the ```network``` shared bindings so to be honest I stopped here.  I suppose I could add something to ```ipaddress``` to further encapsulate the common configuration things (gateway/dns/subnet) but that seemed like a different PR.

File: ports/esp32s2/common-hal/wifi/Radio.c [hunk 3]
```diff
@@ -149,13 +150,66 @@ wifi_radio_error_t common_hal_wifi_radio_connect(wifi_radio_obj_t *self, uint8_t
     return WIFI_RADIO_ERROR_NONE;
 }
 
+mp_obj_t common_hal_wifi_radio_get_ap_info(wifi_radio_obj_t *self) {
+    if (!esp_netif_is_netif_up(self->netif)) {
+        return mp_const_none;
+    }
+
+    // Make sure the interface is in STA mode
+    if (self->sta_mode){
```
Comment (tannewt) [inline ports/esp32s2/common-hal/wifi/Radio.c:159]:
```suggestion
    if (!self->sta_mode){
```

**Review by tannewt — CHANGES_REQUESTED** [review]
One bug I think. Good otherwise.

File: ports/esp32s2/common-hal/wifi/Radio.c (hunk 3 above)
Comment (astrobokonon) [reply ports/esp32s2/common-hal/wifi/Radio.c:159]: You're right!  But this made me wonder why my tests worked at all.  I don't see any reference to anything actually setting ```self->sta_mode``` to True at any point...I'm just double checking to make sure I didn't goof up and remove that somewhere else or if it was just forgotten during ```start_station```.

**Comment by astrobokonon** [comment]
@tannewt The bug you found did actually uncover another one (```self->sta_mode``` was never actually set to true) so I put in a quick fix to at least set ```self->sta_mode``` so it can actually be used.  

Right now nothing will unset it, though.  My guess is that won't really ever occur until either an equivalent ```start_ap``` method is added.

**Review by tannewt — APPROVED** [review]
Haha, oops! Thanks for fixing it up.

**Pull request merged**